File: x10.runtime.cws/src/x10/runtime/cws/Closure.java
/*
 *
 * (C) Copyright IBM Corporation 2007
 *
 *  This file is part of X10 runtime. It is 
 *  governed by the licence under which 
 *  X10 is released.
 *
 */
package x10.runtime.cws;


import static x10.runtime.cws.ClosureStatus.*;

import java.util.concurrent.locks.ReentrantLock;
import java.util.concurrent.locks.Lock;
import java.util.List;
import java.util.ArrayList;


/**
 * A closure corresponds to a slow invocation of a task. This 
 * class is intended to be subclassed by client code. Subclasses
 * may specify extra state (e.g. fields such as result),
 * and will supply implementations of the abstract methods compute
 * and executeAsInlet. The code for a slow invocation of a task
 * should live in compute. The code for executeAsInlet should
 * specify where the result of the closure is to be deposited
 * in the parent closure.
 * 
 * Each worker contains a closure at the top level. When the
 * topmost closure of a victim is stolen, a new child closure is
 * created from the topmost frame in the closure  and left
 * behind on the victim. The old closure is now a parent
 * and is moved to the thief, which starts executing it.
 * 
 * @author vj 04/18/07
 *
 */
public abstract class Closure  {

	protected volatile Cache cache;
	public Frame frame;
	protected Closure parent;
	protected volatile ClosureStatus status;
	protected Lock lock;
	protected Worker lockOwner;
	protected int joinCount;
	
	/**
	 * Inlets are not represented explicitly as separate pieces of code --
	 * they will be once we figure out if we want to support them. (We 
	 * probably do.) Rather we just use a child closure directly as an inlet.
	 */
	protected List<Closure> completeInlets;
	protected List<Closure> incompleteInlets;
	
	protected Worker ownerReadyQueue;
	/**
	 * The ready deque is maintained through a pair of references.
	 */
	protected Closure nextReady, prevReady;
	
	protected ClosureStatus status() { return status;}
	
	protected Closure() {
		super();
		lock = new ReentrantLock();
		initialize();
	}
	
	/**
	 * Returns true if the closure has children
	 * that have not yet joined.
	 * @return
	 */
	protected boolean hasChildren() {
		return joinCount > 0;
	}
	
	
	void lock(Worker agent) { 
		lock.lock();
		lockOwner =agent;
	}
	
	void unlock() { 
		
		lockOwner=null;
		lock.unlock();
	}

	void addCompletedInlet(Closure child) {
		if (completeInlets == null)
			completeInlets = new ArrayList<Closure>();
		completeInlets.add(child);
	}
	

	void makeReady() {
		status=READY;
		cache=null;
	}
	
	void completeAndEnque(Worker ws, Closure child) {
		
		assert (lockOwner == ws);
		addCompletedInlet(child);
		
	}
	
	void removeChild(Closure child) {
		
		if (incompleteInlets != null) 
		 incompleteInlets.remove(child);
		// for (Inlet i : incompleteInlets) {
		//	if (i.target == child) return i;
		//}
		//return null;
	}
	
	/**
	 * This code is executed by the thief on the parent while holding 
	 * the lock on the parent and on the victim.
	 * @param thief  -- The worker performing the steal.
	 * @param victim -- The worker from who work has been stolen.
	 * @return the child closure
	 */
	Closure promoteChild(Worker thief, Worker victim) {
		
		assert lockOwner == thief;
		assert status == RUNNING;
		assert ownerReadyQueue == victim;
		assert victim.lockOwner == thief;
		assert nextReady==null;
		assert cache.head <= cache.exception;

		Frame f = cache.childFrame();
		Closure child = f.makeClosure();
		child.parent = this;
		child.joinCount = 0;
		child.cache = cache;
		child.status = RUNNING;
		child.frame = f;
		child.ownerReadyQueue=null;
		++child.cache.head;
		// TODO: cache.head is never reduced. Hmm. We need to implement
		// wrap around.
		victim.addBottom(thief, child);
		return child;
	}
	/**
	 * This code is executed by the thief on the parent while holding 
	 * the lock on the parent. The lock on the victim has been given up.
	 * Therefore other thiefs may be hitting upon the victim simultaneously.
	 * 
	 * @param thief  -- The worker performing the steal.
	 * @param child -- The child closure being promoted.
	 */
	void finishPromote(Worker thief, Closure child) {
		
		assert lockOwner == thief;
		assert child.lockOwner != thief;
		
		/* Add the child to the parent. */
		++joinCount;
		if (incompleteInlets == null)
			incompleteInlets = new ArrayList<Closure>();
		incompleteInlets.add(child);
		
		/* Set the parent's cache to null and its status to READY */
		makeReady();
		
	}
	/**
	 * Do the thief part of Dekker's protocol.  Return true upon success,
	 * false otherwise.  The protocol fails when the victim already popped
	 * T so that E=T.
	 * Must be the case that Thread.currentThread()==thief.
	 */
	boolean dekker(Worker thief) {
		assert lockOwner==thief;
		
		incrementExceptionPointer(thief);
		Cache c = cache;
		int tail = c.tail;
		int head = c.head;
		if ((head + 1) >= tail) {
			decrementExceptionPointer(thief);
			return false;
		}
		// so there must be at least two elements in the framestack for a theft.
		return true;
	}
    
    private void decrementExceptionPointer(Worker ws) {
    	assert lockOwner == ws;
    	assert status == RUNNING;
    	
    	cache.decrementExceptionPointer();
    }
    
    private void incrementExceptionPointer(Worker ws) {
    	assert lockOwner == ws;
    	assert status == RUNNING;
    	
    	cache.incrementExceptionPointer();
    }
    private void resetExceptionPointer(Worker ws) {
    	 assert lockOwner==ws;
    	 cache.resetExceptionPointer();
     }
    boolean handleException(Worker ws) {
    	resetExceptionPointer(ws);
    	ClosureStatus s = status;
    	assert s == RUNNING || s == RETURNING;
    	//assert (cache.head >= cache.tail);
    	// Other kinds of exceptions are not being handled now.
    	// a steal has happened.
    	status = RETURNING;
    	// The result is going to be moved to the desired location
    	// by application code. 
    	return true;
    	
    }
  
    private void signalImmediateException(Worker ws) {
    	assert lockOwner == ws;
    	assert status == RUNNING;
    	cache.signalImmediateException();
    }
   
    /**
     * Return protocol. The closure has completed its computation. Its return value
     * is now propagated to the parent. The closure to be executed next, possibly null,
     * is returned.  
     * This closure must not be locked (by this worker??) and must not be in any deque.
     * Required that ws==Thread.currentThread().
     * @return the parent, if this is the last child joining.
     */
     private Closure closureReturn(Worker ws) {
    	
    	assert (joinCount==0);
    	assert (status == RETURNING);
    	assert (ownerReadyQueue==null);
    	assert (lockOwner != ws);
    	
    	Closure parent = this.parent;
    	if (parent == null) {
    		// Must be a top level closure.
    		return null;
    	}
    	assert (parent != null);
    	
    	parent.lock(ws);
    	try {
    		assert parent.status != RETURNING;
    		assert parent.frame != null;
    		parent.removeChild(this);
    		lock(ws);
    		parent.completeAndEnque( ws, this);
    		try {
    			if (parent.status == RUNNING) {
    				parent.signalImmediateException(ws);
    			} else
    				parent.pollInlets(ws);
    			// Is a fence() needed?
    			--parent.joinCount;
    			return parent.provablyGoodStealMaybe(ws);
    		} finally {
    			unlock();
    		}
    		
    	} finally {
    		parent.unlock();
    	}
    	// The child should be garbage at this point.
    }
    
    /**
     * Suspend this closure. Called during execution of slow sync
     * when there is at least one outstanding child.
     * ws must be the worker executing suspend.
     * Assume: ws=Thread.currentThread();
     */
   void suspend(Worker ws) {
    	
    	assert lockOwner == ws;
    	assert status == RUNNING;
    	
    	status = SUSPENDED;
    	// throw away the bottommost closure on the worker.
    	// the only references left to this closure are from
    	// its children.
    	Closure cl = ws.extractBottom(ws);
    	assert cl==this;
    	
//    	Setting ownedReadyQueue to null even though Cilk does not do it.
    	cl.ownerReadyQueue=null;
    	
    	
    }
    
    /**
     * Return the parent provided that its joinCount is zero 
     * and it is suspended. The parent should now be considered
     * stolen by the worker that just finished returning the
     * value of a child
     * @return parent or null
     */
    private Closure provablyGoodStealMaybe(Worker ws) {
    	
    	assert lockOwner==ws;
    	assert parent != null;
    	
    	Closure result = null;
    	if (parent.joinCount==0 &&parent.status == SUSPENDED) {
    		result = parent;
    		parent.pollInlets(ws);
    		parent.ownerReadyQueue =null;
    		parent.makeReady();
    	}
    	return result;
    }
	
   
	/**
	 * Run all completed inlets.
	 * TODO: Figure out why pollInlets are being run incrementally
	 * and not just once, after joinCount==0. Perhaps because
	 * this method is supposed to perform abort processing.
	 */
	void pollInlets(Worker ws) {
		
		assert lockOwner==ws;
		
		if (status==RUNNING && ! cache.atTopOfStack()) {
			if (ws.lockOwner !=ws) {
				System.out.println("Cache is " + cache.dump());
			}
			assert ws.lockOwner == ws;
		}
		if (completeInlets != null)
			for (Closure i : completeInlets) {
				i.executeAsInlet();
			}
		completeInlets = null;
	}
	
	 /**
     * Must be called by every slow procedure after it sets the 
     * return value but before it returns.
     * This method ensures that the closure's value is returned
     * to the parent. If this is the last child joining the parent,
     * and the parent is suspended, return the parent (this is a
     * provably good steal).
     * @return
     */
   
    Closure returnValue(Worker ws) {
    	
    	assert status==RETURNING;
    	
    	return closureReturn(ws);
    }
	
	/**
	 * Execute this closure. Performed after the closure has been
	 * stolen from a victim. Will eventually invoke compute(frame),
	 * after setting things up.
	 * @param ws -- the current thread, must be equal to Thread.currentThread()
	 */
	Closure execute(Worker ws) {
		
		assert lockOwner != ws;
		
		lock(ws);
		ClosureStatus s = status;
		Frame f = frame;
		if (s == READY) {
			try {
				status = RUNNING;
		    	// load the cache from the worker's state.
		    	cache = ws.cache;
		    	cache.pushFrame(frame);
		    	cache.resetExceptionPointer();
				assert f != null;
				pollInlets(ws);
			} finally {
				unlock();
			}
			ws.lock(ws);
			try { 
				ws.addBottom(ws, this);
			} finally {
				ws.unlock();
			}
			compute(ws, f);
			return null;
		}
		if (s == RETURNING) {
			unlock();
			return returnValue(ws);
		}
		assert false;
		throw new Error(ws + "executes " + status + " " + this + ": error!");
	}
	
//	=============== The methods intended to be called by client code =======
//	=============== that subclasses Closure.========
	
	/**
	 * Slow execution entry point for the scheduler. Invoked by the thief
	 * running in the scheduler after it has stolen the closure from a victim.
	 * @param w -- The thread invoking the compute, i.e. w == Thread.currentThread()
	 * @param frame -- frame within which to execute
	 */
	abstract protected void compute(Worker w, Frame frame);
    
	/**
	 * Subclasses should override this as appropriate. 
	 * But they should alwas first call super.initialize();
	 *
	 */
	protected void initialize() {
		// need to handle abort processing.
	}

    /* Public method intended to be invoked from within 
     * slow methods of client code.
     * @return true -- iff the closure must suspend because not 
     *                 all children have returned
     */
	final protected boolean sync() {
		Worker ws = ((Worker) Thread.currentThread());
		return ws.sync();
	}
	/**
	 * Invoked by client code before a return from slow code. It will
	 * mark the current closure as RETURNING so it wont be stolen. It will
	 * pop the current frame. 
	 * Before invoking this call, client code is responsible for setting the appropriate state
	 * on the closure so that the value to be returned is known.
	 * 
	 */
	final protected void setupReturn() {
		// Do not trust client code to pass this parameter in.
		Worker ws = (Worker) Thread.currentThread();
		ws.lock(ws);
		try {
			
			Closure t = ws.peekBottom(ws);
			assert t==this;
			lock(ws);
			try {
				assert status==RUNNING;
				status=RETURNING;
				frame=null;
				ws.popFrame();
			} finally {
				unlock();
			}
		} finally {
			ws.unlock();
		}
	}

	
	/**
	 * Return your value to the parent closure. Typically the
	 * closure will be created with information about where
	 * to deposit its result.
	 */
	abstract protected void executeAsInlet();

}

File: x10.lib/src/sched/Closure.cc
/*
============================================================================
 Name        : Closure.cpp
 Author      : Rajkishore Barik
 Version     :
 Copyright   : IBM Corporation 2007
 Description : Exe source file
============================================================================
*/


#include "Closure.h"
#include "Worker.h"
#include "Lock.h"
#include "Cache.h"
#include "Job.h"
#include "Frame.h"
#include "Sys.h"
#include "StealAbort.h"
#include <assert.h>

using namespace std;
using namespace x10lib_cws;

int Closure::getstatus() const { return status;}
Frame *Closure::parentFrame() const { return parent->frame; }
Closure *Closure::getparent() const { return parent; }

// Closure::Closure() { }

Closure::Closure(Frame *frame) {
	done = false;
	this->frame = frame;
	lock_var = new PosixLock();

	//public members to be init
	cache = NULL;
	parent = NULL;
	joinCount = 0;
	nextReady = prevReady = NULL;
	ownerReadyQueue = NULL;
	status = -1; //It is to be set to a valid value by someone; Not me!

	//protected members to be init
	lockOwner = NULL;
	outlet = NULL;
	done = false;

	initialize();
}

Closure::~Closure() { 
	delete lock_var; 
	completeInlets.clear(); 
	incompleteInlets.clear();
	// TODO -- verify these
	//delete frame;
}

bool Closure::hasChildren() const {
	return joinCount > 0;
}
void Closure::lock(Worker *agent) { 
	LOCK(lock_var);
	lockOwner = agent;
}
void Closure::unlock() { 
	lockOwner=NULL;
	UNLOCK(lock_var);
}
void Closure::addCompletedInlet(Closure *child) {
		/*if (completeInlets == NULL)
			completeInlets = new list<Closure>();*/
		completeInlets.push_back(child);
}
	
void Closure::removeChild(Closure *child) {
		//if (incompleteInlets != NULL) 
		 incompleteInlets.remove(child);
		// for (Inlet i : incompleteInlets) {
		//	if (i.target == child) return i;
		//}
		//return NULL;
}
	
	/**
	 * This code is executed by the thief on the parent while holding 
	 * the lock on the parent and on the victim.
	 * @param thief  -- The worker performing the steal.
	 * @param victim -- The worker from who work has been stolen.
	 * @return the child closure
	 */
Closure *Closure::promoteChild(Worker *thief, Worker *victim) {
		
		assert(lockOwner == thief);
		assert(status == RUNNING);
		assert(ownerReadyQueue == victim);
		assert(victim->lockOwner == thief);
		assert(nextReady==NULL);
		assert(cache->exceptionOutstanding());

		Frame *childFrame = cache->childFrame();
		Closure *child = childFrame->makeClosure();
		
		Frame *pFrame = cache->topFrame();
		pFrame->setOutletOn(child);
		
		// Leave the parent link in there.
		// It will not be used by globally quiescent computations.
		child->parent = this;
		child->joinCount = 0;
		child->cache = cache;
		child->status = RUNNING;
		child->ownerReadyQueue = NULL;
		cache->incHead();
		
		victim->addBottom(thief, child);
		return child;
}
	/**
	 * This code is executed by the thief on the parent while holding 
	 * the lock on the parent. The lock on the victim has been given up.
	 * Therefore other thiefs may be hitting upon the victim simultaneously.
	 * 
	 * @param thief  -- The worker performing the steal.
	 * @param child -- The child closure being promoted.
	 */
void Closure::finishPromote(Worker *thief, Closure *child) {
		
		assert(lockOwner == thief);
		assert(child == NULL || child->lockOwner != thief);
		
		/* Add the child to the parent. */
		if (! child->requiresGlobalQuiescence())
			++joinCount;
		// No need to add child to parent's incomplete inlets
		// unless aborts are being propagated.
		//if (incompleteInlets == NULL)
		//	incompleteInlets = new ArrayList<Closure>();
		//incompleteInlets.add(child);
		
		/* Set the parent's cache to NULL and its status to READY */
		status=READY;
		cache=NULL;
}

	/**
	 * Do the thief part of Dekker's protocol.  Return true upon success,
	 * false otherwise.  The protocol fails when the victim already popped
	 * T so that E=T.
	 * Must be the case that Thread.currentThread()==thief.
	 */
bool Closure::dekker(Worker *thief) {
		assert(lockOwner==thief);
		assert(status==RUNNING);
		return cache->dekker(thief);
}
    
/*
void Closure::decrementExceptionPointer(Worker *ws) {
    	assert(lockOwner == ws);
    	assert(status == RUNNING);
    	cache->decrementExceptionPointer();
}
    
void Closure::incrementExceptionPointer(Worker *ws) {
    	assert(lockOwner == ws);
    	assert(status == RUNNING);
    	
    	cache->incrementExceptionPointer();
}
void Closure::resetExceptionPointer(Worker *ws) {
    	 assert(lockOwner==ws);
    	 
    	 cache->resetExceptionPointer();
}
    
*/
bool Closure::handleException(Worker *ws) {
    	cache->resetExceptionPointer(ws);
    	
    	assert (status == RUNNING || status == RETURNING);
    	if (cache->empty()) {
    		assert(joinCount==0);
    		status = RETURNING;
        	return true;
    	}
    	return false;
    	
}
  
void Closure::signalImmediateException(Worker *ws) {
    	assert(lockOwner == ws);
    	assert(status == RUNNING);
    	cache->signalImmediateException();
}



 /**
     * Return protocol. The closure has completed its computation. Its return value
     * is now propagated to the parent. The closure to be executed next, possibly NULL,
     * is returned.  
     * This closure must not be locked (by this worker??) and must not be in any deque.
     * Required that ws==Thread.currentThread().
     * @return the parent, if this is the last child joining.
     */
Closure *Closure::closureReturn(Worker *w) {
    	
    	// Short circuit for globally quiescent computations.
    	if (requiresGlobalQuiescence() && parent != NULL) {
    		w->currentJob()->accumulateResultInt(resultInt());
    		return NULL;
    	}
    	assert (joinCount==0);
    	assert (ownerReadyQueue==NULL);
    	assert (lockOwner != w);
    	
    	if (! requiresGlobalQuiescence())
    		completed();
    	Closure *parent = this->parent;
    	if (parent == NULL) {
    		// Must be a top level closure.
    		return NULL;
    	}
    	return parent->acceptChild(w, this);
}
     
Closure *Closure::acceptChild(Worker *ws, Closure *child) {
		Closure *res;
    	lock(ws);
    	assert(status != RETURNING);
    	assert(frame != NULL);
    		//removeChild(child);
    	assert (joinCount > 0); // ADDED by RAJ;
    	MEM_BARRIER();
    	--joinCount;
    	MEM_BARRIER();
    	child->lock(ws);
    	assert (lockOwner == ws);
    	addCompletedInlet(child);
    	res = provablyGoodStealMaybe(ws, child);
    	child->unlock();
    	unlock();
    	// TODO RAJ: Please free up child now -- its no longer needed
    	return res;
    	// The child should be garbage at this point.
}
    
    /**
     * Suspend this closure. Called during execution of slow sync
     * when there is at least one outstanding child.
     * ws must be the worker executing suspend.
     * Assume: ws=Thread.currentThread();
     */
void Closure::suspend(Worker *ws) {
    	assert(lockOwner == ws);
    	assert(status == RUNNING);
    	
    	status = SUSPENDED;
    	
    	// throw away the bottommost closure on the worker.
    	// the only references left to this closure are from
    	// its children.
    	Closure *cl = ws->extractBottom(ws);
    	assert(cl==this);
    	
//    	Setting ownedReadyQueue to NULL even though Cilk does not do it.
    	cl->ownerReadyQueue=NULL;
}
    
    /**
     * Return the parent provided that its joinCount is zero 
     * and it is suspended. The parent should now be considered
     * stolen by the worker that just finished returning the
     * value of a child
     * @return parent or NULL
     */
Closure *Closure::provablyGoodStealMaybe(Worker *ws, Closure *child) {
    	assert(child->lockOwner==ws);
    	//assert parent != NULL;
    	Closure *result = NULL;
    	
    	if (joinCount==0 && status == SUSPENDED) {
    		result = this;
    		pollInlets(ws);
    		ownerReadyQueue = NULL;
    		status=READY;
    		cache=NULL;
    		/*if (Worker.reporting) {
    			System.out.println(ws + " awakens " + this);
    		}*/
    	} 
    	return result;
    }
	
   
	/**
	 * Run all completed inlets.
	 * TODO: Figure out why pollInlets are being run incrementally
	 * and not just once, after joinCount==0. Perhaps because
	 * this method is supposed to perform abort processing.
	 */
void Closure::pollInlets(Worker *ws) {
		list<Closure *>::iterator i;
		assert(lockOwner==ws);
		
		if (status==RUNNING && ! cache->atTopOfStack()) {
			assert(ws->lockOwner == ws);
		}
		if (!completeInlets.empty())
			/* TODO LIST TRAVERSAL -- RAJ*/
			for (i=completeInlets.begin(); i != completeInlets.end(); ++i) 
				(*i)->executeAsInlet();
			
		completeInlets.clear();
}

	
	 /**
     * Must be called by every slow procedure after it sets the 
     * return value but before it returns.
     * This method ensures that the closure's value is returned
     * to the parent. If this is the last child joining the parent,
     * and the parent is suspended, return the parent (this is a
     * provably good steal).
     * @return
     */
   
Closure *Closure::returnValue(Worker *ws) {
    	assert(status==RETURNING);
    	
    	return closureReturn(ws);
}
	
	/**
	 * Execute this closure. Performed after the closure has been
	 * stolen from a victim. Will eventually invoke compute(frame),
	 * after setting things up.
	 * @param w -- the current thread, must be equal to Thread.currentThread()
	 */
Closure *Closure::execute(Worker *w) {
		Closure *res = NULL;
		assert(lockOwner != w);
		
		lock(w);
		
		Frame *f = frame;
		switch (status) {
		case READY:	
			status = RUNNING;
		    // load the cache from the worker's state.
		    cache = w->cache;
		    cache->pushFrame(frame);
		    cache->resetExceptionPointer(w);
			assert(f != NULL);
			pollInlets(w);
			
			unlock();
			
			w->lock(w);
			w->addBottom(w, this);
			w->unlock();
			
			try {
				compute(w, f); 
			}
			catch(StealAbort *saEx) {
				//Nothing much to do. It just unwound the stack as we wanted
				delete saEx; 
			}
			res = NULL;
			break;
		
		case RETURNING:
			unlock();
			res = returnValue(w);
			break;
		default:
			abort(); // TODO RAJ
			//throw new Error(w + "executes " + status + " " + this + ": error!");
		}
		return res;
}

	/**
	 * Return your value to the parent closure. Typically the
	 * closure will be created with information about where
	 * to deposit its result.
	 */
void Closure::executeAsInlet() {
		if (outlet != NULL) {
			outlet->run();
		}
}
//	=============== The methods intended to be called by client code =======
//	=============== that subclasses Closure.========
	
	/**
	 * Slow execution entry point for the scheduler. Invoked by the thief
	 * running in the scheduler after it has stolen the closure from a victim.
	 * @param w -- The thread invoking the compute, i.e. w == Thread.currentThread()
	 * @param frame -- frame within which to execute
	 */
void Closure::compute(Worker *w, Frame *frame) {abort();}
    
	/**
	 * Subclasses should override this as appropriate. 
	 * But they should alwas first call super.initialize();
	 *
	 */
void Closure::initialize() {
		// need to handle abort processing.
}

    /* Public method intended to be invoked from within 
     * slow methods of client code.
     * @return true -- iff the closure must suspend because not 
     *                 all children have returned
     */
bool Closure::sync(Worker *ws) {
		return ws->sync();
}
	
	
	/**
	 * Invoked by client code before a return from slow code. It will
	 * mark the current closure as RETURNING so it wont be stolen. It will
	 * pop the current frame. 
	 * Before invoking this call, client code is responsible for setting the appropriate state
	 * on the closure so that the value to be returned is known.
	 * 
	 */
void Closure::setupReturn(Worker *w) {
		// Do not trust client code to pass this parameter in.
		//Worker *w = (Worker) Thread.currentThread(); // TODO how to handle this RAJ
		done = true;
		w->lock(w);
		
		if (requiresGlobalQuiescence()) {
			w->extractBottom(w);
			// speed the result on its way.
			if (parent != NULL)
			w->currentJob()->accumulateResultInt(resultInt());
			return;
		}
		Closure *t = w->peekBottom(w);
		assert(t==this);
		lock(w);
			
		assert(status==RUNNING);
		status=RETURNING;
		//frame=NULL;
		w->popFrame();
			
		unlock();
			
		w->unlock();
}

void Closure::setupGQReturnNoArg(Worker *w) {
		// Do not trust client code to pass this parameter in.
		//Worker *w = (Worker) Thread.currentThread(); //TODO RAJ
		w->lock(w); 
		w->extractBottom(w);
		w->popFrame();
		w->unlock();
		
}

void Closure::setupGQReturn(Worker *w) {
		// Do not trust client code to pass this parameter in.
		//Worker *w = (Worker) Thread.currentThread(); // TODO RAJ
		// do not set done to true. This will be 
		// done when global quiescence is recognized.
		w->lock(w);
		Closure *t = w->peekBottom(w);
		assert(t==this);
		lock(w);
		assert(status==RUNNING);
		status=RETURNING;
		//frame=NULL;
		w->popFrame();
		unlock();
		w->unlock();
		
}
	
	
void Closure::setOutlet(Outlet *o) { outlet=o;}
/** Replace the frame by a copy. 
	 * 
	 * <p> Called during
	 * a steal to ensure that frames are not shared between the caches
	 * of two different workers. Must be called with the current thread,
	 * w, holding the lock on the victim. This ensures that the 
	 * victim is not running freely, modifying the frame being
	 * copied. Why? Because the Theft Protocol ensures that the
	 * victim cannot return to processing the frame until it has checked
	 * whether this frame has been stolen. If the frame has been stolen,
	 * the victim must grab the lock on itself. So therefore we must
	 * ensure that this method is called by the thief before it releases 
	 * the lock on the victim. 
	 * 
	 * Note that the victim is by definition the owner of this closure.
	 * ownerReadyQueue must not be null.
	 *
	 */
void Closure::copyFrame(Worker *w) {
		assert(ownerReadyQueue->lockOwner==w);
		frame = frame->copy();
}	
bool Closure::isDone() const { return done;}
	
//RuntimeException Closure::getException() { return NULL;} // TODO RAJ
	
	/**
	 * Invoked on completion of the computation associated with this closure. 
	 * May be overridden by client code. Note: This method may be invoked more than once
	 * by the scheduler.
	 *
	 */
void Closure::completed() volatile {
		done = true;
}
	
	// At most one of the following pairs of methods should be overridden by subclassing
	// closures. No pair may be overridden if the closure does not have an associated
	// return value. These methods are not abstract so that Closure can be used directly
	// when there is no reason to subclass it.
void Closure::setResultInt(int x) { abort(); }
void Closure::accumulateResultInt(int x) { abort();}
int Closure::resultInt() { abort(); return 0;}
	
void Closure::setResultFloat(float x) {abort();}
void accumulateResultFloat(float x) { abort();}
float Closure::resultFloat() { abort(); return 0.0;}
	
void Closure::setResultLong(long x) {abort();}
void Closure::accumulateResultLong(long x) { abort();}
long Closure::resultLong() { abort(); return 0l;}
	
void Closure::setResultDouble(double x) {abort();}
void Closure::accumulateResultDouble(double x) { abort();}
double Closure::resultDouble() { abort(); return 0.0;}
	
void Closure::setResultObject(void *x) {abort();}
void *Closure::resultObject() { abort(); return NULL; }
bool Closure::requiresGlobalQuiescence() volatile { abort(); return false; }
